File: app/src/main/java/ru/nekotyan2d/mirea_switcher/data/repository/AccountRepository.kt
package ru.nekotyan2d.mirea_switcher.data.repository

import android.content.Context
import androidx.core.content.edit
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import ru.nekotyan2d.mirea_switcher.data.model.Account

class AccountRepository(ctx: Context) {
    private val prefs = ctx.getSharedPreferences("accounts", Context.MODE_PRIVATE)
    private val gson = Gson()
    private val type = object : TypeToken<MutableList<Account>>() {}.type

    fun getAll(): MutableList<Account> {
        return gson.fromJson(prefs.getString("accounts", "[]"), type)
    }

    fun save(accounts: List<Account>) {
        prefs.edit { putString("accounts", gson.toJson(accounts)) }
    }

    fun addIfAbsent(token: String): Boolean {
        val accounts = getAll()
        if (accounts.any { it.token == token }) return false
        accounts.add(Account("Account ${accounts.size + 1}", token))
        save(accounts)
        return true
    }
}

File: lib/data/repository/account_repository.dart
import 'dart:convert';
import 'package:mirea_switcher/data/models/account.dart';
import 'package:shared_preferences/shared_preferences.dart';

class AccountRepository {
  static const _key = 'accounts';
  final SharedPreferences _prefs;

  AccountRepository(this._prefs);

  List<Account> getAll() {
    final raw = _prefs.getString(_key) ?? '[]';
    final list = jsonDecode(raw) as List<dynamic>;
    return list.map((e) => Account.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<void> save(List<Account> accounts) async {
    final encoded = jsonEncode(accounts.map((a) => a.toJson()).toList());
    await _prefs.setString(_key, encoded);
  }

  Future<bool> addIfAbsent(String name, String token) async {
    final accounts = getAll();
    if(accounts.any((a) => a.token == token)) return false;
    accounts.add(Account(name: name, token: token));
    await save(accounts);
    return true;
  }

  Future<bool> remove(String token) async {
    final accounts = getAll()..removeWhere((a) => a.token == token);
    await save(accounts);
    return true;
  }
}